### PR TITLE
Feature/rewrite address api

### PR DESCRIPTION
**Pull request by Young666666Jian**
# PR Details

Change implementation of getting txId
## Description

In original code, transaction id is got by traversing whole state.  This version use transaction type and address to get transaction id directly
## Related Issue
#113 

## Motivation and Context
Increase performance of address api

## How Has This Been Tested
Unit test has passed

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the code style of this project.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/main/scala/vsys/api/http/TransactionsApiRoute.scala [hunk 1]
```diff
@@ -172,15 +172,35 @@ case class TransactionsApiRoute(
     pathPrefix(Segment) { address =>
       Address.fromString(address) match {
         case Left(e) => complete(ApiError.fromValidationError(e))
-        case Right(a) =>
-          complete(Json.arr(JsArray(state.activeLeases().flatMap(state.transactionInfo)
+        case Right(a) => {
+          if(stateSettings.txTypeAccountTxIds){
+            val txNum = state.txTypeAccTxLengths(TransactionType.LeaseTransaction, a)
+            val txIds = state.txTypeAccountTxIds(TransactionType.LeaseTransaction, a, txNum, 0)
+            if (txIds._2 != null && txIds._2.size > 0) {
```
Comment (ncying) [inline src/main/scala/vsys/api/http/TransactionsApiRoute.scala]: avoid to use null in scala

File: src/main/scala/vsys/api/http/TransactionsApiRoute.scala [hunk 2]
```diff
@@ -172,15 +172,35 @@ case class TransactionsApiRoute(
...
+                .filter(a => state.isLeaseActive(a._3.asInstanceOf[LeaseTransaction]))
+                .collect{
+                  case (h:Int, tx:ProcessedTransaction, lt:LeaseTransaction)
+                    if EllipticCurve25519Proof.fromBytes(lt.proofs.proofs.head.bytes.arr).toOption.get.publicKey.address == address =>
+                    processedTxToExtendedJson(tx) + ("height" -> JsNumber(h))
+                }
+              )
+            res match {
+              case Some(tx) => complete(Json.arr(JsArray(tx)))
+              case None => complete(StatusCodes.NotFound -> Json.obj("message" -> "Address has no transaction"))
```
Comment (zl730) [inline src/main/scala/vsys/api/http/TransactionsApiRoute.scala]: no txs does not mean NotFound

File: src/main/scala/vsys/api/http/TransactionsApiRoute.scala [hunk 3]
```diff
@@ -172,15 +172,35 @@ case class TransactionsApiRoute(
     pathPrefix(Segment) { address =>
       Address.fromString(address) match {
         case Left(e) => complete(ApiError.fromValidationError(e))
-        case Right(a) =>
-          complete(Json.arr(JsArray(state.activeLeases().flatMap(state.transactionInfo)
+        case Right(a) => {
+          if(stateSettings.txTypeAccountTxIds){
+            val txNum = state.txTypeAccTxLengths(TransactionType.LeaseTransaction, a)
+            val txIds = state.txTypeAccountTxIds(TransactionType.LeaseTransaction, a, txNum, 0)
+            val res = Option(txIds._2)
```
Comment (zl730) [inline src/main/scala/vsys/api/http/TransactionsApiRoute.scala]: Option is useless

File: src/main/scala/vsys/api/http/TransactionsApiRoute.scala [hunk 4]
```diff
@@ -172,15 +172,35 @@ case class TransactionsApiRoute(
...
+              .map(_.flatMap(state.transactionInfo)
```
Comment (zl730) [inline src/main/scala/vsys/api/http/TransactionsApiRoute.scala]: ????

File: src/main/scala/vsys/api/http/TransactionsApiRoute.scala [hunk 5]
```diff
@@ -172,15 +172,35 @@ case class TransactionsApiRoute(
...
+                .map(a => (a._1,a._2,a._2.transaction))
```
Comment (zl730) [inline src/main/scala/vsys/api/http/TransactionsApiRoute.scala]: do not copy the codes below.  result of `state.activeLeases()` is already filtered by lease status

File: src/main/scala/vsys/api/http/TransactionsApiRoute.scala [hunk 6]
```diff
@@ -172,15 +172,35 @@ case class TransactionsApiRoute(
...
+                .filter(a => state.isLeaseActive(a._3.asInstanceOf[LeaseTransaction]))
+                .collect{
+                  case (h:Int, tx:ProcessedTransaction, lt:LeaseTransaction)
+                    if EllipticCurve25519Proof.fromBytes(lt.proofs.proofs.head.bytes.arr).toOption.get.publicKey.address == address =>
```
Comment (zl730) [inline src/main/scala/vsys/api/http/TransactionsApiRoute.scala]: useless filter

File: src/main/scala/vsys/api/http/TransactionsApiRoute.scala [hunk 7]
```diff
@@ -172,15 +172,27 @@ case class TransactionsApiRoute(
...
         case Left(e) => complete(ApiError.fromValidationError(e))
-        case Right(a) =>
-          complete(Json.arr(JsArray(state.activeLeases().flatMap(state.transactionInfo)
-              .map(a => (a._1,a._2,a._2.transaction))
+        case Right(a) => {
+          if(stateSettings.txTypeAccountTxIds){
+            val txNum = state.txTypeAccTxLengths(TransactionType.LeaseTransaction, a)
+            val txIds = state.txTypeAccountTxIds(TransactionType.LeaseTransaction, a, txNum, 0)
+            complete(Json.arr(JsArray(txIds._2
+              .flatMap(state.transactionInfo)
```
Comment (zl730) [inline src/main/scala/vsys/api/http/TransactionsApiRoute.scala]: `txIds._2.flatMap(state.transactionInfo)` is already implemented in `txTypeAccountTransactions`

File: src/main/scala/vsys/api/http/TransactionsApiRoute.scala [hunk 8]
```diff
@@ -172,15 +172,27 @@ case class TransactionsApiRoute(
...
+              .map(a => (a._1, a._2, a._2.transaction))
+              .filter(a => state.isLeaseActive(a._3.asInstanceOf[LeaseTransaction]))
+              .map(a => (processedTxToExtendedJson(a._2) + ("height" -> JsNumber(a._1))))
```
Comment (zl730) [inline src/main/scala/vsys/api/http/TransactionsApiRoute.scala]: can refactored to one `.map`

File: src/main/scala/vsys/api/http/TransactionsApiRoute.scala [hunk 9]
```diff
@@ -172,15 +172,27 @@ case class TransactionsApiRoute(
...
-          complete(Json.arr(JsArray(state.activeLeases().flatMap(state.transactionInfo)
-              .map(a => (a._1,a._2,a._2.transaction))
+        case Right(a) => {
+          if(stateSettings.txTypeAccountTxIds){
+            val txNum = state.txTypeAccTxLengths(TransactionType.LeaseTransaction, a)
+            val txIds = state.txTypeAccountTxIds(TransactionType.LeaseTransaction, a, txNum, 0)
+            complete(Json.arr(JsArray(txIds._2
+              .flatMap(state.transactionInfo)
+              .map(a => (a._1, a._2, a._2.transaction))
+              .filter(a => state.isLeaseActive(a._3.asInstanceOf[LeaseTransaction]))
```
Comment (zl730) [inline src/main/scala/vsys/api/http/TransactionsApiRoute.scala]: call `isLeaseActive` in loop is not in one transaction. need read lock here

**Pull request closed without merging**